kotlin/workflow-runtime/src/main/java/com/squareup/workflow/debugging/RealWorkflowDebugger.kt

```diff
@@ -0,0 +1,90 @@
+package com.squareup.workflow.debugging
+
+import com.squareup.workflow.diagnostic.WorkflowDiagnosticListener
+import com.squareup.workflow.diagnostic.andThen
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.ConflatedBroadcastChannel
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.asFlow
+
+internal class RealWorkflowDebugger : WorkflowDebugger {
+
+  @Volatile override var emitTrees: Boolean = false
+
+  @Volatile override var breakpointEnabled: Boolean = false
+
+  var diagnosticListener: WorkflowDiagnosticListener? = null
+    private set
+
+  private var isRunning = false
+
+  private val resumeSignal = Channel<Unit>(capacity = 1)
```

Comment:
Wouldn't a `RendezvousChannel` be appropriate here since the usage should exclusive be "one runtimewaits, then one debugger resumes". It would also allow using the return value of `offer` to indicate whether the debugger did actually resume.

Reply:
It's got a buffer to avoid a race condition in the scenario where a collector of `trees` calls `resume` before we actually start receiving on the resume signal. In that case, the resume call would get dropped. The API is less complicated to reason about this way, because by the time the tree is emitted, any `resume` call will actually resume the runtime. Do you have a use case for `resume` returning whether it actually succeeded or not? I could see it maybe being useful to expose a property indicating whether we're paused or not, but that would be exposed as its own Flow, not a return value from `resume`.